File: packages/tool-sdk/src/receipt.ts
import type {
  PolicyDecision,
  ResultType,
  ToolAttemptStatus,
  ToolRiskLevel,
  VerificationRecord,
} from "@semi-os/shared";

export interface ToolTarget {
  kind: string;
  displayName: string;
  canonicalReference?: string;
}

export interface ApprovalBinding {
  approvalId: string;
  payloadDigest: string;
  approvedAtMs: number;
}

export interface ArtifactReference {
  artifactId: string;
  kind: string;
  uri: string;
  digest?: string;
}

/**
 * 工具执行的审计回执。
 *
 * `inputSummary` 必须脱敏，`inputDigest` 用于绑定实际规范化输入。成功状态
 * 仍必须由 `verification.status === "passed"` 支撑，不能只相信工具返回文本。
 */
export interface ToolExecutionReceipt {
  attemptId: string;
  taskId: string;
  toolName: string;
  idempotencyKey: string;
  attemptNumber: number;
  status: ToolAttemptStatus;
  riskLevel: ToolRiskLevel;
  inputSummary: string;
  inputDigest: string;
  target?: ToolTarget;
  policyDecision: PolicyDecision;
  approval?: ApprovalBinding;
  startedAtMs: number;
  finishedAtMs?: number;
  durationMs?: number;
  resultType?: ResultType;
  outputSummary?: string;
  verification?: VerificationRecord;
  artifacts: readonly ArtifactReference[];
}

/**
 * 检查工具执行是否为已验证的成功。
 *
 * 成功必须同时满足：状态为 succeeded、结果类型为 success、验证状态为 passed。
 */
export function isVerifiedSuccess(receipt: ToolExecutionReceipt): boolean {
  return (
    receipt.status === "succeeded" &&
    receipt.resultType === "success" &&
    receipt.verification?.status === "passed"
  );
}

File: crates/protocol/src/receipt.rs
use crate::{
    PolicyDecision, ResultType, ToolAttemptStatus, ToolRiskLevel, VerificationRecord,
    VerificationStatus,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolTarget {
    pub kind: String,
    pub display_name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub canonical_reference: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ApprovalBinding {
    pub approval_id: String,
    pub payload_digest: String,
    pub approved_at_ms: i64,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArtifactReference {
    pub artifact_id: String,
    pub kind: String,
    pub uri: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub digest: Option<String>,
}

/// 跨 Worker/Host 边界的工具执行回执。
///
/// `input_summary` 只能包含脱敏后的可读摘要，规范化原始输入通过摘要绑定。
/// `Succeeded` 同时要求 `ResultType::Success` 和明确通过的验证记录。
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ToolExecutionReceipt {
    pub attempt_id: String,
    pub task_id: String,
    pub tool_name: String,
    pub idempotency_key: String,
    pub attempt_number: i64,
    pub status: ToolAttemptStatus,
    pub risk_level: ToolRiskLevel,
    pub input_summary: String,
    pub input_digest: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target: Option<ToolTarget>,
    pub policy_decision: PolicyDecision,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub approval: Option<ApprovalBinding>,
    pub started_at_ms: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub finished_at_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub duration_ms: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result_type: Option<ResultType>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_summary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub verification: Option<VerificationRecord>,
    pub artifacts: Vec<ArtifactReference>,
}

impl ToolExecutionReceipt {
    pub fn is_verified_success(&self) -> bool {
        self.status == ToolAttemptStatus::Succeeded
            && self.result_type == Some(ResultType::Success)
            && self
                .verification
                .as_ref()
                .is_some_and(|record| record.status == VerificationStatus::Passed)
    }
}
